### PR TITLE
Ensure error handler always restored

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
@demeritcowboy I think this will address https://github.com/civicrm/civicrm-core/pull/30068

Before
----------------------------------------
error handler not restored on exception

After
----------------------------------------
now it is

Technical Details
----------------------------------------
Re-organization of try-catch with addition of finally

Comments
----------------------------------------

### CODE REVIEW

File: CRM/Utils/String.php [hunk 1]
```diff
@@ -1035,30 +1035,32 @@ public static function parseOneOffStringThroughSmarty($templateString) {
...
-          'template' => $templateString,
-        ]);
-        throw new \CRM_Core_Exception('Message was not parsed due to invalid smarty syntax : ' . $e->getMessage() . ((CIVICRM_UF === 'UnitTest' || CRM_Utils_Constant::value('SMARTY_DEBUG_STRINGS')) ? $templateString : ''));
+      else {
+        $templateString = (string) $smarty->fetch('string:{eval var=$smartySingleUseString|smarty:nodefaults}');
       }
     }
-    else {
-      $templateString = (string) $smarty->fetch('string:{eval var=$smartySingleUseString|smarty:nodefaults}');
+    catch (SmartyCompilerException $e) {
```
Comment (demeritcowboy) [inline CRM/Utils/String.php:1053]: I agree with the concept of adding "finally" but I think this exception doesn't exist in smarty2 which explains why it was in the `if`. Also it seemed to be throwing a RuntimeException in the test which was why it wasn't getting caught here.

Comment (eileenmcnaughton) [reply CRM/Utils/String.php:1053]: oh....

Comment (eileenmcnaughton) [reply CRM/Utils/String.php:1053]: what say we just catch Exception?

Comment (demeritcowboy) [reply CRM/Utils/String.php:1053]: Sounds good.

**Comment by eileenmcnaughton** [comment]
@demeritcowboy interestingly over & above the tax_amount ones which I'm close to resolving this PR also surfaced notices in message templates (although I'm not close on figuring those out)

**Comment by demeritcowboy** [comment]
Yeah my thinking was they might be fallout from earlier fails. I think you're referring to the selfsvc ones?
For the message template form one: https://github.com/civicrm/civicrm-core/pull/30070

**Comment by demeritcowboy** [comment]
let's see where we're at - jenkins retest this please

**Comment by eileenmcnaughton** [comment]
So at least one of those is in here I think https://github.com/civicrm/civicrm-core/pull/30079 & I think you have one open for Messagetemplates admin

**Comment by demeritcowboy** [comment]
jenkins retest this please

**Comment by eileenmcnaughton** [comment]
Just one left -  I did have a poke at this earlier so I will fix


CRM_Financial_Form_PaymentFormsTest::testEventPaymentForms
Undefined array key "tax_amount"

/home/homer/buildkit/build/build-1/web/sites/all/modules/civicrm/CRM/Price/BAO/PriceField.php:786
/home/homer/buildkit/build/build-1/web/sites/all/modules/civicrm/CRM/Price/BAO/PriceField.php:363
/home/homer/buildkit/build/build-1/web/sites/all/modules/civicrm/ext/eventcart/CRM/Event/Cart/Form/Checkout/ParticipantsAndPrices.php:96
/home/homer/buildkit/build/build-1/web/sites/all/modules/civicrm/ext/eventcart/CRM/Event/Cart/Form/Checkout/ParticipantsAndPrices.php:27
/home/homer/buildkit/build/build-1/web/sites/all/modules/civicrm/tests/phpunit/CRM/Financial/Form/PaymentFormsTest.php:99
/home/homer/buildkit/build/build-1/web/sites/all/modules/civicrm/tests/phpunit/CiviTest/CiviUnitTestCase.php:284
/home/homer/buildkit/extern/phpunit9/phpunit9.phar:2307

**Comment by demeritcowboy** [comment]
```
CRM_Financial_Form_PaymentFormsTest::testEventPaymentForms
Creation of dynamic property CRM_Event_Cart_Form_Checkout_Payment::$_price_values is deprecated

/home/homer/buildkit/build/build-4/web/sites/all/modules/civicrm/ext/eventcart/CRM/Event/Cart/Form/Checkout/Payment.php:183
```

Thanks for working on this. Hopefully it won't get too deep into eventcart. For what it's worth if I just declare the above variable then a couple more come up, and then if I just blindly declare those it works.

  public $_price_values;
  public $_paymentFields;
  public $sub_trxn_index;
  public $trxn_id;
  public $trxn_date;

**Comment by demeritcowboy** [comment]
yay!

**Comment by eileenmcnaughton** [comment]
thanks for finding this issue @demeritcowboy  - it was hiding problems from us

**Pull request merged**